### PR TITLE
Introduce PostgreSQL support

### PR DESCRIPTION
**Pull request by m110**
This PR introduces support for PostgreSQL databases. It's based on the same principles as the MySQL implementation.

Summary:
* `github.com/lib/pq` used in tests as the DB driver
* Tested with PostgreSQL 11.5
* Renamed `DefaultSchema` to `DefaultMySQLSchema`. The old name can still be used (it's a type alias), but is now deprecated.
* Changed `NextOffsetQuery` to not use `MAX`, but `COALESCE` with subquery instead. `FOR UPDATE` doesn't work with aggregate queries in Postgres.
* A few cosmetic changes.

What's left:
- [ ] Documentation
- [ ] Godoc review 
- [ ] Add examples

### CODE REVIEW

File: pkg/sql/offsets_adapter.go [hunk 1]
```diff
@@ -32,45 +34,102 @@ type DefaultMySQLOffsetsAdapter struct {
...
-func (s DefaultMySQLOffsetsAdapter) ConsumedMessageQuery(
+func (a DefaultMySQLOffsetsAdapter) ConsumedMessageQuery(
 	topic string,
 	offset int,
 	consumerGroup string,
 	consumerULID []byte,
 ) (string, []interface{}) {
-	// offset_consumed is not queried anywhere, it's used only to detect race conditions with NextOffsetQuery.
-	ackQuery := `INSERT INTO ` + s.MessagesOffsetsTable(topic) + ` (offset_consumed, consumer_group) 
+	// offset_consumed is not queried anywhere, it'a used only to detect race conditions with NextOffsetQuery.
```
Comment (m110) [inline pkg/sql/offsets_adapter.go]:
```suggestion
	// offset_consumed is not queried anywhere, it's used only to detect race conditions with NextOffsetQuery.
```

File: pkg/sql/offsets_adapter.go [hunk 2]
```diff
@@ -32,45 +34,102 @@ type DefaultMySQLOffsetsAdapter struct {
...
+	return fmt.Sprintf(`"watermill_offsets_%s"`, topic)
+}
+
+func (a DefaultPostgresOffsetsAdapter) ConsumedMessageQuery(
+	topic string,
+	offset int,
+	consumerGroup string,
+	consumerULID []byte,
+) (string, []interface{}) {
+	// offset_consumed is not queried anywhere, it'a used only to detect race conditions with NextOffsetQuery.
```
Comment (m110) [inline pkg/sql/offsets_adapter.go]:
```suggestion
	// offset_consumed is not queried anywhere, it's used only to detect race conditions with NextOffsetQuery.
```

File: pkg/sql/offsets_adapter.go [hunk 3]
```diff
@@ -32,45 +34,102 @@ type DefaultMySQLOffsetsAdapter struct {
...
 	consumerULID []byte,
 ) (string, []interface{}) {
 	// offset_consumed is not queried anywhere, it's used only to detect race conditions with NextOffsetQuery.
-	ackQuery := `INSERT INTO ` + s.MessagesOffsetsTable(topic) + ` (offset_consumed, consumer_group) 
+	ackQuery := `INSERT INTO ` + a.MessagesOffsetsTable(topic) + ` (offset_consumed, consumer_group)
 		VALUES (?, ?) ON DUPLICATE KEY UPDATE offset_consumed=VALUES(offset_consumed)`
+	return ackQuery, []interface{}{offset, consumerGroup}
+}
 
+// DefaultPostgresOffsetsAdapter is adapter for storing offsets in PostgreSQL database.
```
Comment (roblaszczak) [inline pkg/sql/offsets_adapter.go]: maybe separated file + rename?

File: pkg/sql/schema_adapter.go [hunk 4]
```diff
@@ -74,69 +117,100 @@ func (s DefaultSchema) SchemaInitializingQueries(topic string) []string {
...
+func (s DefaultMySQLSchema) MessagesTable(topic string) string {
+	if s.GenerateMessagesTableName != nil {
+		return s.GenerateMessagesTableName(topic)
 	}
+	return fmt.Sprintf("`watermill_%s`", topic)
+}
 
-	msg = message.NewMessage(string(r.UUID), r.Payload)
+// DefaultPostgresSchema is a default implementation of SchemaAdapter based on PostgreSQL.
+type DefaultPostgresSchema struct {
```
Comment (roblaszczak) [inline pkg/sql/schema_adapter.go]: to be concrete, it probably should be `DefaultPostgreSQLSchema` :poultry_leg:

File: pkg/sql/schema_adapter.go [hunk 5]
```diff
@@ -74,69 +117,100 @@ func (s DefaultSchema) SchemaInitializingQueries(topic string) []string {
...
+	// GenerateMessagesTableName may be used to override how the messages table name is generated.
```
Comment (roblaszczak) [inline pkg/sql/schema_adapter.go]: and maybe separated file too (for MySQL also)

File: pkg/sql/schema_adapter.go [hunk 6]
```diff
@@ -74,69 +117,100 @@ func (s DefaultSchema) SchemaInitializingQueries(topic string) []string {
...
-	if r.Metadata != nil {
-		err = json.Unmarshal(r.Metadata, &msg.Metadata)
-		if err != nil {
-			return 0, nil, errors.Wrap(err, "could not unmarshal metadata as JSON")
-		}
+func (s DefaultPostgresSchema) SchemaInitializingQueries(topic string) []string {
+	createMessagesTable := strings.Join([]string{
+		`CREATE TABLE IF NOT EXISTS ` + s.MessagesTable(topic) + ` (`,
+		`"offset" SERIAL,`,
+		`"uuid" VARCHAR(36) NOT NULL,`,
```
Comment (roblaszczak) [inline pkg/sql/schema_adapter.go]: doesn't it have built-in uuid type?

File: pkg/sql/schema_adapter_test.go [hunk 7]
```diff
@@ -1,19 +1,83 @@
...
 
+	"github.com/stretchr/testify/require"
+
+	"github.com/ThreeDotsLabs/watermill"
 	"github.com/ThreeDotsLabs/watermill-sql/pkg/sql"
+	"github.com/ThreeDotsLabs/watermill/message"
 )
 
-// testSchema makes the following changes to DefaultSchema to comply with tests:
+func TestDefaultMySQLSchema(t *testing.T) {
```
Comment (roblaszczak) [inline pkg/sql/schema_adapter_test.go]: it would be nice to have some comment, explaining what this check is checking and why ;)

Comment (roblaszczak) [reply pkg/sql/schema_adapter_test.go]: just to not forget

File: pkg/sql/schema_adapter.go (hunk 5 above)
Comment (m110) [reply pkg/sql/schema_adapter.go]: See 4e5b4d04ba067d42d000aa3054848a0cce24f7ad

File: pkg/sql/schema_adapter.go (hunk 4 above)
Comment (m110) [reply pkg/sql/schema_adapter.go]: Renamed all related types to PostgreSQL, see 4e5b4d04ba067d42d000aa3054848a0cce24f7ad

File: pkg/sql/offsets_adapter.go (hunk 3 above)
Comment (m110) [reply pkg/sql/offsets_adapter.go]: See 4e5b4d04ba067d42d000aa3054848a0cce24f7ad

File: pkg/sql/schema_adapter.go (hunk 6 above)
Comment (m110) [reply pkg/sql/schema_adapter.go]: It does, although it's used for UUID only, so it's not that universal (won't accept an ULID for example).

File: pkg/sql/schema_adapter_test.go (hunk 7 above)
Comment (m110) [reply pkg/sql/schema_adapter_test.go]: See cc53b9e2bd237d45bd0fd44441b787a6a311dcab

**Pull request merged**